java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/PushDownRefactoring.java

```diff
@@ -0,0 +1,223 @@
+                continue;
+            }
+            if (m instanceof TypeElement && types.isSubtype(m.asType(), element.asType())) {
+                continue;
+            }
+            TreePath mPath = trees.getPath(m);
+            long startMember = mPath != null ? sourcePositions.getStartPosition(mPath.getCompilationUnit(), mPath.getLeaf()) : -1;
+            long endMember = mPath != null ? sourcePositions.getEndPosition(mPath.getCompilationUnit(), mPath.getLeaf()) : -1;
+            boolean selected = offset > startMember && offset < endMember;
+            members.add(new QuickPickItem(createLabel(info, m), null, null, selected, new ElementData(m)));
+        }
+        if (members.isEmpty()) {
+            return Collections.emptyList();
+        }
+        QuickPickItem elementItem = new QuickPickItem(createLabel(info, element));
+        elementItem.setUserData(new ElementData(element));
+        return Collections.singletonList(createCodeAction(Bundle.DN_PushDown(), PUSH_DOWN_REFACTORING_KIND, null, PUSH_DOWN_REFACTORING_COMMAND, uri, elementItem, members));
+    }
+
+    @Override
+    public Set<String> getCommands() {
+        return commands;
+    }
+
+    @Override
+    @NbBundle.Messages({
+        "DN_SelectMembersToPushDown=Select members to push down",
+    })
+    public CompletableFuture<Object> processCommand(NbCodeLanguageClient client, String command, List<Object> arguments) {
+        try {
+            if (arguments.size() > 2) {
+                String uri = gson.fromJson(gson.toJson(arguments.get(0)), String.class);
+                QuickPickItem sourceItem = gson.fromJson(gson.toJson(arguments.get(1)), QuickPickItem.class);
+                List<QuickPickItem> members = Arrays.asList(gson.fromJson(gson.toJson(arguments.get(2)), QuickPickItem[].class));
+                client.showQuickPick(new ShowQuickPickParams(Bundle.DN_SelectMembersToPushDown(), true, members)).thenAccept(selected -> {
+                    if (selected != null && !selected.isEmpty()) {
+                        pushDown(client, uri, sourceItem, selected);
+                    }
+                });
+            } else {
+                throw new IllegalArgumentException(String.format("Illegal number of arguments received for command: %s", command));
+            }
+        } catch (Exception ex) {
+            client.showMessage(new MessageParams(MessageType.Error, ex.getLocalizedMessage()));
+        }
+        return CompletableFuture.completedFuture(true);
+    }
+
+    private void pushDown(NbCodeLanguageClient client, String uri, QuickPickItem source, List<QuickPickItem> members) {
+        try {
+            FileObject file = Utils.fromUri(uri);
+            ClasspathInfo info = ClasspathInfo.create(file);
+            JavaSource js = JavaSource.forFileObject(file);
+            if (js == null) {
+                throw new IOException("Cannot get JavaSource for: " + uri);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [IOException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [183].  Consider adding a catch clause of this specific exception in the outer try statement.